pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,314 @@
+package expr
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+func ConvertToLong(frames data.Frames) (data.Frames, error) {
+	if len(frames) == 0 {
+		// general empty case for now
+		return frames, nil
+	}
+	// Four Conversion Possible Cases
+	// 1. NumericMulti -> NumericLong
+	// 2. NumericWide -> NumericLong
+	// 3. TimeSeriesMulti -> TimeSeriesLong
+	// 4. TimeSeriesWide -> TimeSeriesLong
+
+	// Detect if input type is declared
+	// First Check Frame Meta Type
+
+	var inputType data.FrameType
+	if frames[0].Meta != nil && frames[0].Meta.Type != "" {
+		inputType = frames[0].Meta.Type
+	}
+	if inputType != "" && !supportedToLongConversion(inputType) {
+		return frames, fmt.Errorf("unsupported input type %s for SQL expression", inputType)
+	}
+
+	// TODO: Add some guessing of Type if not declared
+
+	if inputType == "" {
+		return frames, fmt.Errorf("could not determine input type")
+	}
+
+	convert := getToLongConversionFunc(inputType)
+	if convert == nil {
+		return frames, fmt.Errorf("could not get conversion function for input type %s", inputType)
+	}
+
+	return convert(frames)
+}
+
+func convertNumericMultiToNumericLong(frames data.Frames) (data.Frames, error) {
+	// Apart from metadata, NumericMulti is basically NumericWide, except one frame per thing
+	// so we collapse into wide and call the wide conversion
+	wide := convertNumericMultiToNumericWide(frames)
+	return convertNumericWideToNumericLong(wide)
+}
+
+func convertNumericMultiToNumericWide(frames data.Frames) data.Frames {
+	newFrame := data.NewFrame("")
+	for _, frame := range frames {
+		for _, field := range frame.Fields {
+			if !field.Type().Numeric() {
+				continue
+			}
+			newField := data.NewFieldFromFieldType(field.Type(), field.Len())
+			newField.Name = field.Name
+			newField.Labels = field.Labels.Copy()
+			for i := 0; i < field.Len(); i++ {
+				newField.Set(i, field.CopyAt(i))
+			}
```

Comment:
We're doing this zero or one times, right? I wonder if an `if` check would communicate this more clearly than a `for` loop